Updated ENTRYPOINT to exec syntax for Windows Dockerfile to workaround containerd issue

michaelongso opened the pull request:
# Background

[SC-48697](https://app.shortcut.com/octopusdeploy/story/48697/issue-with-windows-tentacle-container-running-in-fargate)

We have a reported issue where Windows Tentacle Container fails to run in ECS Fargate. The issue is linked back to the open [containerd issue](https://github.com/containerd/containerd/issues/5067) when parsing the _shell_ form for `ENTRYPOINT` syntax. The known workaround is to use the [_exec_ form](https://docs.docker.com/engine/reference/builder/#understand-how-cmd-and-entrypoint-interact:~:text=has%20two%20forms%3A-,The%20exec%20form,-%2C%20which%20is%20the).

# Results

<!-- Describe the result of the change -->

Fixes [issue-8143](https://github.com/OctopusDeploy/Issues/issues/8143)

# Behavioural Update from _shell_ to _exec_

The interaction between `ENTRYPOINT` and `CMD` is now changed due to the usage of _exec_ syntax, where additional `CMD` will now be appended as part of the `ENTRYPOINT` arguments. Previously with _shell_ syntax, additional `CMD` is ignored.

See the [docker doc](https://docs.docker.com/engine/reference/builder/#understand-how-cmd-and-entrypoint-interact:~:text=Understand%20how%20CMD%20and%20ENTRYPOINT%20interact%F0%9F%94%97) for a details explanation.

## Before

Running Windows Tentacle Container on ECS Fargate will result in:
```
CannotStartContainerError: CannotStartContainerError: hcs::System::CreateProcess: "cmd /S /C powershell -File ./configure-tentacle.ps1 -Verbose && powershell -File ./run-tentacle.ps1 -Verbose" 0353f39e1acb446bbf43d99137706834-1111712579: The system cannot find the file specified.: unknown
```

## After

![Screenshot 2023-06-09 at 12 30 03 pm](https://github.com/OctopusDeploy/OctopusTentacle/assets/15998611/48aed78d-5852-49cd-b3f2-a87627ea28d8)

![Screenshot 2023-06-09 at 12 30 51 pm](https://github.com/OctopusDeploy/OctopusTentacle/assets/15998611/d22ab073-9a80-422b-990a-75dae7a92608)

## Validation with Azure Container with Windows OS

![Screenshot 2023-06-09 at 12 31 03 pm](https://github.com/OctopusDeploy/OctopusTentacle/assets/15998611/467b4f9c-126e-41c8-b5fb-f59d9b54b48a)


# How to review this PR

<!--
Describe how you want people to review the pull request.
Perhaps you just want an "in principal" review to prove an idea.
Perhaps you want specific people to test the resulting changes.
-->

Quality :heavy_check_mark:
<!-- Describe focus areas (if any): Review tests/ Exploratory testing/ Smoke testing? -->

# Pre-requisites

- [x] I have read [How we use GitHub Issues](https://github.com/OctopusDeploy/Issues/blob/master/docs/CONTRIBUTING.internal.md) for help deciding when and where it's appropriate to make an issue.
- [x] I have considered informing or consulting the right people, according to the [ownership map](https://whimsical.com/ownership-map-NzbiD4HJyvhC9jNJNfS6TG).
- [x] I have considered appropriate testing for my change.

geofflamrock commented on docker/windows/Scripts/configure-and-run-tentacle-wrapper.ps1:
```diff
@@ -0,0 +1,2 @@
+﻿& ./configure-tentacle.ps1 -Verbose
```
Am I right in remembering that the `&` here makes the script run in the background?

LukeButters replied:
in powershell it seems `&` at the start means "call a process" and `&` at the end means run in the background. How confusing!
https://ss64.com/ps/call.html
https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_operators?view=powershell-7.3&viewFallbackFrom=powershell-7.1#call-operator-
https://stackoverflow.com/questions/22074507/what-does-the-symbol-in-powershell-mean#:~:text=%26%20is%20the%20call%20operator%20which,a%20script%2C%20or%20a%20function.

geofflamrock replied:
Ah right, yeah that could take a bit to get used to. My potential concern was the two scripts not running in sequence anymore, but seems fine if the `&` at the start just means "run this process and wait please".

LukeButters replied:
The tentacle tests do run the windows tentacle docker container, so if it is completely broken the test chain should catch it.

geofflamrock reviewed (approved):
LGTM, a couple of things you might consider doing based on when we chatted:
- The PR title has `CMD` in it which was the original approach we were looking to take, it might need updating now
- We identified that having `ENTRYPOINT` in exec form changes the behaviour if someone were to run this container with a specific command. Previously this would be ignored and now it will be appended to the command. Because we are calling a script now I'm not sure whether this will fail or not. It feels like it could be worth noting this in the PR description in case anyone comes across it.

Another thought: What does the `ENTRYPOINT` for the linux container look like? Is it worth ensuring these are consistent?

michaelongso commented:
> The PR title has CMD
Updated.

> We identified that having ENTRYPOINT in exec form changes the behaviour

Added _ Behavioural Update from shell to exec_ section in the PR description.

> Another thought: What does the ENTRYPOINT for the linux container look like
The `Dockerfile` for the Linux container is using `CMD`. There's little commonality due to the requirement for `.sh` and `.ps1`, so it is not worthwhile to attempt consistency.
